src/manipulation.js

```diff
@@ -134,38 +134,50 @@ jQuery.extend( {
 	},
 
 	cleanData: function( elems ) {
-		var data, elem, type,
-			special = jQuery.event.special,
-			i = 0;
-
-		for ( ; ( elem = elems[ i ] ) !== undefined; i++ ) {
-			if ( acceptData( elem ) ) {
-				if ( ( data = elem[ dataPriv.expando ] ) ) {
-					if ( data.events ) {
-						for ( type in data.events ) {
-							if ( special[ type ] ) {
-								jQuery.event.remove( elem, type );
-
-							// This is a shortcut to avoid jQuery.event.remove's overhead
-							} else {
-								jQuery.removeEvent( elem, type, data.handle );
-							}
-						}
-					}
+	    var data, elem, type,
+	        special = jQuery.event.special,
+	        i = 0;
+
+	    // Convert elems to a native array if it is not already
+	    elems = jQuery.makeArray(elems);
+
+	    while (i < elems.length) {
+	        elem = elems[i];
+
+	        if (acceptData(elem)) {
+	            if ((data = elem[dataPriv.expando])) {
+	                if (data.events) {
+	                    for (type in data.events) {
+	                        if (special[type]) {
+	                            jQuery.event.remove(elem, type);
+
+	                        // This is a shortcut to avoid jQuery.event.remove's overhead
+	                        } else {
+	                            jQuery.removeEvent(elem, type, data.handle);
+	                        }
+	                    }
+	                }
+
+	                // Support: Chrome <=35 - 45+
+	                // Assign undefined instead of using delete, see Data#remove
+	                elem[dataPriv.expando] = undefined;
+	            }
+	            if (elem[dataUser.expando]) {
+
+	                // Support: Chrome <=35 - 45+
+	                // Assign undefined instead of using delete, see Data#remove
+	                elem[dataUser.expando] = undefined;
+	            }
+
+	            // Remove the element from the array while preserving the index positions
+	            elems.splice(i, 1);
```

Comment:
For large `elems` arrays, this could be very expensive. I also don't see why we'd need to do that.

Reply:
@mgol  reason was to preserve the index position while deletion